### ISSUE TITLE
How to use action masking?

### ISSUE THREAD

**Issue by Zebin-Li** [tensorforce/tensorforce#792]
Hi!
I was trying to use action masking, however, I cannot find some very good examples, so could you please mention some resources here? Thank you very very much!

Best,
Zebin LI

**Comment by AlexKuhnle** [comment]
Hi @Zebin-Li,

does [this docs example](https://tensorforce.readthedocs.io/en/latest/basics/features.html#action-masking) help? Basically, you add an extra `[ACTION-NAME]_mask` value to the state, with a boolean mask of which of the `num_values` actions are available at each step. You can either add this to the state the environment returns before passing it on to the agent, or preferably your environment returns a state with the mask values included, in which case you can, for instance, simply use the `Runner` utility as usual. Let me know if there are any questions.

**Comment by Zebin-Li** [comment]
Thank you@AlexKuhnle for the quick response!

Actually, I read the docs example and I have some questions about it. 
1. In the example, action_mask = [True, False, True], does it mean the first and the third action are valid, or the second action is valid? In other words, "True" means applying action mask on the action or this action is valid?
2. Is the action mask editable during the running? In my case, my action is to choose one image from a pool, and if the image has been chosen, it should not be chosen again. For example, at the beginning I have 50 images so the action space is 50 (the actions can be represented by indexes), after the first timestep, one image is chosen and the action space is 49 now. I may want to use the action mask to invalidate that already chosen action. Therefore, during this process, the action mask is changing. Is it possible to do that?

Thank you again!
Zebin Li

**Comment by AlexKuhnle** [comment]
1. Good question: True means this action is valid/available, False means that it is masked/not available. I think that's how masking is commonly implemented, e.g. in NumPy.
2. The action mask can change at every timestep. So in your case, the "environment" can keep track of what actions have been taken in this episode already (input to `execute()`), and return the corresponding action mask value as part of the state at every timestep (output of `execute()`).

Hope that helps!

**Comment by Zebin-Li** [comment]
Thank you for the nice help!
I have another question that do I need to do something about "enable_int_action_masking"? Cause I saw it when I searching "action mask" and there is a question about it [https://github.com/tensorforce/tensorforce/issues/775](url). I am using PPO, but there is no "enable_int_action_masking" option in the list of parameters of PPO. Therefore, I am not sure whether it is related to the implementation of action mask. 
Thank you!

**Comment by AlexKuhnle** [comment]
The docs for this argument are [here](https://tensorforce.readthedocs.io/en/latest/agents/tensorforce.html) (search "enable_int_action_masking"), note the comment at the end of the [argument list for PPO](https://tensorforce.readthedocs.io/en/latest/agents/ppo.html). You can set it via `config=dict(enable_int_action_masking=True)`, however, this is not necessary since it's true by default. The idea is that if you don't want to use it and you want a more minimal agent computation graph, you can turn this feature off and with it the generation of related operations in the computation graph.

**Comment by Zebin-Li** [comment]
Thank you very very much! I will try it!

**Comment by Zebin-Li** [comment]
Hi!
I have a question that where should I define the action_mask? I first defined it in the environment like:
![image](https://user-images.githubusercontent.com/52910998/119395240-392dee00-bca1-11eb-8304-859b099303d1.png)
And I create the environment and the agent like:
![image](https://user-images.githubusercontent.com/52910998/119395446-8316d400-bca1-11eb-9d6b-472b553887ff.png)
But I get an error like:
![image](https://user-images.githubusercontent.com/52910998/119395572-b48f9f80-bca1-11eb-8d01-75e727eedf50.png)

Then I tried to define the action_mask when creating the agent like:
![image](https://user-images.githubusercontent.com/52910998/119395754-fae4fe80-bca1-11eb-8571-208a1999334e.png)
![image](https://user-images.githubusercontent.com/52910998/119395789-06d0c080-bca2-11eb-9dc7-4ca37996b50e.png)
But I get an error like:
![image](https://user-images.githubusercontent.com/52910998/119395869-1d771780-bca2-11eb-83ef-c5e72769b4de.png)

So where is the correct location or what is the correct way to define the action_mask? Thank you very much!

Best,
Zebin Li

**Comment by AlexKuhnle** [comment]
Few things:
- `Environment.states()` only should to return the specification of the actual state space, whereas the action-masks can automatically be inferred from the action space.
- `Agent.create()` argument `states` should only be passed the state space spec, so `states=environment.states()` as usual.
- The action masks are passed for `agent.act()` and can be returned as part of `Environment.reset/execute()`, so very roughly:

```python

class MyEnv(Environment):

    def reset(self):
        ...
        state_plus_mask = {'state': state, 'action_mask': ???}
        return state_plus_mask

    def execute(self, actions):
        ...
        state_plus_mask = {'state': state, 'action_mask': ???}
        return state_plus_mask, terminal, reward

states = environment.reset()  # contains both "state" and "action_mask"
actions = agent.act(states)  # "action_mask" determines masking, can change for every timestep
```

**Comment by Zebin-Li** [comment]
Thank you very much!
However, another issue occurred...
what I modified for the code is:

    `class CustomEnv(Environment):

    def __init__(self, ...):
        super().__init__()
        ...
        self.choose_one_mask = [True] * 10

    def compute_timestep(self, action): 
        idx = action["choose_one"]
        queried_image = copy.deepcopy(self.action_space[idx, :])
        return idx, queried_image

    def actions(self):
        return {
            "choose_one": dict(type="int", num_values = 10)
        }

    def reset(self):
        ...
        self.choose_one_mask = [True]*10
        state_plus_mask = {"state":state, "choose_one_mask": self.choose_one_mask}
        return state_plus_mask

    def execute(self, actions):
        ...
        idx, queried_image = self.compute_timestep(actions)
        self.choose_one_mask[idx] = 'False'
        state_plus_mask = {"state": state, "choose_one_mask": self.choose_one_mask}
        return state_plus_mask, terminal, reward

    actions = agent.act(states=states)
    states, terminal, reward = environment.execute(actions=actions)`

I get this error:
![image](https://user-images.githubusercontent.com/52910998/119416085-dcdac680-bcc0-11eb-958a-56e861921b6e.png)

In the execute(), after I get an idx, which actually is an action, I want to turn off it by the mask[idx] = False.
I am not sure why the above error happened...
Thank you again!

**Comment by AlexKuhnle** [comment]
Your screenshot of the exception has some of the message cut off... However, I assume the problem might be that you set `self.choose_one_mask[idx] = 'False'`, where `'False'` is a string, not a boolean. Removing the `'` might solve the problem. Otherwise could you post the full exception message?

**Comment by Zebin-Li** [comment]
Hi @AlexKuhnle !
The full screenshot is like:
![image](https://user-images.githubusercontent.com/52910998/119559540-d146d880-bd70-11eb-9d71-a54597036f15.png)
![image](https://user-images.githubusercontent.com/52910998/119559602-e459a880-bd70-11eb-87aa-80e556ac748b.png)
And I changed the 'False' to False, but still get the same error.
Thank you!

**Comment by AlexKuhnle** [comment]
Can you print the `states` value returned by `environment.reset()` / `environment.execute()`, before passing it to `agent.act(...)`? This value definitely looks far bigger than the 10 entries the value should have according to your code.

**Comment by Zebin-Li** [comment]
Ohhh!!! Thank you so much! And sorry for the misleading. Actually, the size of the state is the size of the image pool + model_accuracy, and the size of the action is the same as the image pool. The above code I wrote 10 is just for simplifying, since the size of the image pool is varying (due to randomly split the dataset) and a little large. I just checked and I found I made a mistake that the size of the action mask should be equal to that of the action, not the state. Now I will run it on my experiment code to check the model performance. Thank you again! And I really appreciate your patience and time!

Best,
Zebin Li
